trieste/space.py

```diff
@@ -976,10 +957,102 @@ def get_subspace(self, tag: str) -> SearchSpace:
 
+    def subspace_sample(self, num_samples: int, seed: Optional[int] = None) -> Sequence[TensorType]:
+        """
+        Sample randomly from the space by sampling from each subspace
+        and returning the resulting samples in the same order as specified when initializing
+        the space.
+
+        :param num_samples: The number of points to sample from each subspace.
+        :param seed: Optional tf.random seed.
+        :return: ``num_samples`` i.i.d. random points, sampled uniformly,
+            from each search subspace with shape '[num_samples, D]' , where D is the search space
+            dimension.
+        """
+        tf.debugging.assert_non_negative(num_samples)
+        if seed is not None:  # ensure reproducibility
+            tf.random.set_seed(seed)
+        return [self._spaces[tag].sample(num_samples, seed=seed) for tag in self._tags]
+
+    def __eq__(self, other: object) -> bool:
+        """
+        :param other: A search space.
+        :return: Whether the search space is identical to this one.
+        """
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self._tags == other._tags and self._spaces == other._spaces
+
+
+class TaggedProductSearchSpace(CollectionSearchSpace):
+    r"""
+    Product :class:`SearchSpace` consisting of a product of
+    multiple :class:`SearchSpace`. This class provides functionality for
+    accessing either the resulting combined search space or each individual space.
+
+    Note that this class assumes that individual points in product spaces are
+    represented with their inputs in the same order as specified when initializing
+    the space.
+    """
+
+    def __init__(self, spaces: Sequence[SearchSpace], tags: Optional[Sequence[str]] = None):
+        r"""
+        Build a :class:`TaggedProductSearchSpace` from a list ``spaces`` of other spaces. If
+        ``tags`` are provided then they form the identifiers of the subspaces, otherwise the
+        subspaces are labelled numerically.
+
+        :param spaces: A sequence of :class:`SearchSpace` objects representing the space's subspaces
+        :param tags: An optional list of tags giving the unique identifiers of
+            the space's subspaces.
+        :raise ValueError (or tf.errors.InvalidArgumentError): If ``spaces`` has a different
+            length to ``tags`` when ``tags`` is provided or if ``tags`` contains duplicates.
+        """
+
+        super().__init__(spaces, tags)
+        subspace_sizes = self.subspace_dimension
+
+        self._subspace_sizes_by_tag = {
+            tag: subspace_size for tag, subspace_size in zip(self._tags, subspace_sizes)
+        }
+
+        self._subspace_starting_indices = dict(
+            zip(self._tags, tf.cumsum(subspace_sizes, exclusive=True))
+        )
+
+        self._dimension = tf.cast(tf.reduce_sum(subspace_sizes), dtype=tf.int32)
+
+    @property
+    @check_shapes("return: [D]")
```

Comment:
How much work would it be to specify this in the SearchSpace API and use inherit_shapes wherever appropriate (i.e. everywhere apart from TaggedMultiSearchSpace)? Ditto for sample and dimension, the last of which should presumably be trivial as it's the same everywhere.

Reply:
It should have been simple. So I gave it a go, but have discovered that `@inherit_check_shapes` doesn't seem to work with `@property` methods, whereas `@check_shapes` does. This requires further investigation, so I think we should leave this to a separate PR.